wasm-bindgen-rayon and new OMT plugin

RReverser opened the pull request:
Few months ago, I've extracted and published universal Rust + Rayon adapter as a separate library: https://github.com/GoogleChromeLabs/wasm-bindgen-rayon. It means we no longer need to support our own custom glue code in Squoosh, but can use it directly as a dependency, which is what this PR does.

However, in the process of working on wasm-bindgen-rayon I also made a series of changes to make it work with rollup-plugin-off-main-thread. Those changes made it closer to real AMD, as well as reduced size of the generated code, made it simpler etc., but, unfortunately, turned out to be incompatible with some of the tricks we use in Squoosh - in particular, defining multiple modules inline in a single HTML page.

Previously this worked because OMT was replacing Rollup-generated anonymous `define(` calls with named entry chunks, but that caused other problems when using multiple AMD modules on the same page, so I've removed that on the OMT side and made all the `define`s anonymous again. To preserve current behaviour on Squoosh side, I copied the updated loader code, and exposed an internal `nextDefineUri` as a global variable. This way, build system can emit `nextDefineUri="..."` before each `define(` call, which, in turn, will allow multiple defines on the same page again. This is a hack, but it works and preserves existing behaviour without having to extract everything into separate scripts as Rollup normally expects.

RReverser committed:
WIP: wasm-bindgen-rayon and new OMT plugin

RReverser committed:
Bump package-lock

RReverser committed:
Make OMT work again

RReverser committed:
Update year

RReverser committed:
Restore accidental change

RReverser commented on lib/client-bundle-plugin.js:
```diff
@@ -137,10 +137,24 @@ export default function (inputOptions, outputOptions, resolveFileUrl) {
+        const inlineDefines = [
+          ...allModules.map(
+            (item) =>
+              `nextDefineUri=location.origin+${resolveFileUrl(item)};${
```
@jakearchibald Quick question: here I'm just calling `resolveFileUrl(item)` since our `resolveFileUrl` only needs `fileName` property anyway, but I see that other calls in the same file use

```js
        return resolveFileUrl({
          fileName: clientEntry.fileName,
          moduleId,
          format,
        });
```

or

```js
            return resolveFileUrl({
              fileName: entry.fileName,
              moduleId,
              format: outputOptions.format,
            });
```

instead of just passing object directly.

Should I do the same?

jakearchibald replied:
I think it's fine. If we change format there'll be a lot of refactoring anyway.

RReverser commented on src/features/encoders/oxiPNG/worker/oxipngEncode.ts:
```diff
@@ -77,7 +40,7 @@ export default async function encode(
 ): Promise<ArrayBuffer> {
   if (!wasmReady) {
-    wasmReady = (await threads()) ? initMT() : initST();
+    wasmReady = threads().then((hasThreads: boolean) => hasThreads ? initMT() : initST());
```
Note: this is unrelated fix, just noticed that we weren't assiging Promise immediately, which could result in bugs.

RReverser commented on rollup.config.js:
```diff
@@ -37,13 +37,6 @@ function resolveFileUrl({ fileName }) {
 
-// With AMD output, Rollup always uses document.baseURI, which breaks in workers.
-// This fixes it:
-function resolveImportMeta(property, { chunkId }) {
```
This is no longer necessary because OMT plugin takes care of `import.meta.url` directly now.

jakearchibald replied:
so good!

RReverser commented on lib/css-plugin.js:
```diff
@@ -48,7 +48,7 @@ const appendCssSource = `
 `;
 
-export default function (resolveFileUrl) {
+export default function () {
```
This param was unused so I just removed it.

RReverser commented on codecs/oxipng/build.sh:
```diff
@@ -6,6 +6,4 @@ rm -rf pkg,{-parallel}
 wasm-pack build -t web
 RUSTFLAGS='-C target-feature=+atomics,+bulk-memory' wasm-pack build -t web -d pkg-parallel -- -Z build-std=panic_abort,std --features=parallel
-# Workaround https://github.com/rustwasm/wasm-bindgen/issues/2133:
-sed -i "s|maybe_memory:|maybe_memory?:|" pkg-parallel/squoosh_oxipng.d.ts
```
Also unrelated fix: I've fixed typings upstream in wasm-bindgen, so this hack is no longer necessary.

surma commented:
I’m not sure what broke, but staging version doesn’t work :D

RReverser commented:
Hm, it worked locally, let me recheck.

(If only we had tests... :P)

RReverser commented:
Heh, it's something that I first assumed wouldn't work, but then tried with `npm run dev` and it did, so I thought it's fine, but apparently it fails only in optimized builds. Fixing...

RReverser commented:
Ah no, that's some minifier issue where it doesn't see one of the variable usages. Sigh.

RReverser committed:
Prevent minification of `nextDefineUri`

RReverser commented:
"Fixed".

RReverser committed:
Merge branch 'dev' into wasm-bindgen-rayon

jakearchibald reviewed (changes requested):
There's something not quite right here. With the current version of squoosh.app we ensure that the HTML contains module definitions for everything needed for the first interaction, so we can get to first interaction without external dependencies.

However, this version introduces https://deploy-preview-1007--squoosh.netlify.app/c/logo-with-text.svg-8d769ff2.js as an external module needed for first-interaction.

The HTML contains the definition for this module, but it doesn't appear to be used.

jakearchibald commented on lib/client-bundle-plugin.js:
```diff
@@ -137,10 +137,24 @@ export default function (inputOptions, outputOptions, resolveFileUrl) {
+        const inlineDefines = [
+          ...allModules.map(
+            (item) =>
+              `self.nextDefineUri=location.origin+${resolveFileUrl(item)};${
```
I know we aren't governed by https://url.spec.whatwg.org/#goals, but maybe worth considering:

> Standardize on the term URL. URI and IRI are just confusing. In practice a single algorithm is used for both so keeping them distinct is not helping anyone.

RReverser replied:
> Hmm, good catch. Maybe it's not correctly defined then, let me recheck.

Hah, right, it's actually something I correctly handled first, but then reverted because app seemed to work fine anyway. Fixed.

RReverser replied:
I guess this makes sense; however the right place to change it would be upstream in https://github.com/surma/rollup-plugin-off-main-thread/blob/master/loader.ejs first, as this name comes from there (and I want to keep Squoosh loader not _too_ different from upstream one to make future updates easier).

jakearchibald commented:
Additionally, when I `npm run dev`, and load the result in a browser, I get `(index):59 Uncaught TypeError: depsNames.map is not a function`

Edit: I hadn't ran `npm i`, oops

RReverser commented:
> However, this version introduces [deploy-preview-1007--squoosh.netlify.app/c/logo-with-text.svg-8d769ff2.js](https://deploy-preview-1007--squoosh.netlify.app/c/logo-with-text.svg-8d769ff2.js) as an external module needed for first-interaction.

Hmm, good catch. Maybe it's not correctly defined then, let me recheck.

RReverser committed:
Delay loading external deps

This gives inline `define` calls a chance to define dependencies for earlier `define` calls on the same page.

RReverser commented:
> Hmm, good catch. Maybe it's not correctly defined then, let me recheck.

Hah, right, it's actually something I correctly handled first, but then reverted because app seemed to work fine anyway. Fixed.

jakearchibald commented:
I could also put the dependencies first in the HTML, but I'm a little worried that dependencies-of-dependencies might go wrong.

jakearchibald commented on lib/omt.ejs:
```diff
@@ -13,80 +13,61 @@
+      exports,
+      require
+    };
+    registry[uri] = Promise.resolve().then(() => Promise.all(depsNames.map(
```
Comment, so we don't delete this in future 😄

RReverser replied:
Done.

RReverser committed:
Add comment

jakearchibald reviewed (approved):
Love it. I know this was a pain in the arse. I wish there was an easier, standard way to inline modules.

RReverser commented:
> I wish there was an easier, standard way to inline modules.

In theory, Rollup should really take care of this - as in, one should be able to say "if module size is < given threshold, inline it like you would if it wasn't code-split".

surma commented:
> I wish there was an easier, standard way to inline modules.

OH DO I HAVE SOMETHING FOR YOU

RReverser commented:
> OH DO I HAVE SOMETHING FOR YOU

WELL DO YOU?!

surma commented:
Needed to finish the post :D

https://github.com/littledan/proposal-module-fragments/issues/5#issuecomment-840562415
